File: lib/test/t_NamedDAG_std.cxx
#include <openturns/OT.hxx>

#include "otagrum/NamedDAG.hxx"
#include <iostream>

void testConstructor()
{
  std::cout << "\n\n** Checking constructor " << std::endl;

  const auto proto = "A->C->B->D;E->A->B<-E";
  const auto bn = gum::BayesNet<double>::fastPrototype(proto);

  std::cout << "      proto : " << proto << std::endl;
  std::cout << "         BN : " << bn << std::endl << std::endl;

  OTAGRUM::NamedDAG ndag(bn);
  ndag.setName("ndagFromTest");

  std::cout << "       size : " << ndag.getSize() << std::endl;
  std::cout << "       desc : " << ndag.getDescription() << std::endl
            << std::endl;

  for (gum::NodeId nod = 0; nod < ndag.getSize(); nod++)
  {
    std::cout << " parents(" << nod << ") : " << ndag.getParents(nod)
              << std::endl;
    std::cout << "children(" << nod << ") : " << ndag.getChildren(nod)
              << std::endl;
  }

  std::cout << "topologicalOrder : " << ndag.getTopologicalOrder() << std::endl;

  std::cout << std::endl << ndag.__str__("__str__ | ") << std::endl;

  std::cout << std::endl << "todo : " << ndag.toDot();
}

void testSerialisation()
{
  std::cout << "\n\n** Checking serialization " << std::endl;
  const auto proto = "A->C->B->D;E->A->B<-E";
  const auto bn = gum::BayesNet<double>::fastPrototype(proto);
  OTAGRUM::NamedDAG ndag(bn);
  ndag.setName("ndagFromTest");

  std::cout << "Before saving, " << ndag.getClassName() << ", "
            << ndag.getName() << std::endl;
  OT::Log::Show(OT::Log::ALL);
  OT::Study study("tmp.xml");
  study.add("ndag", ndag);
  study.save();

  std::cout << "\n------------------\n";
  OT::Study study2("tmp.xml");
  study2.load();
  std::cout << "Labels=" << study2.printLabels() << std::endl;
  std::cout << "\n------------------\n";

  OTAGRUM::NamedDAG dag;
  std::cout << "Before fill, " << dag.getClassName() << ", " << dag.getName()
            << std::endl;

  study2.fillObject("ndag", dag);
  std::cout << "After fill, " << dag.getClassName() << ", " << dag.getName()
            << std::endl;

  std::cout << "        size : " << dag.getSize() << std::endl;
  std::cout << "       desc : " << dag.getDescription() << std::endl
            << std::endl;

  for (gum::NodeId nod = 0; nod < ndag.getSize(); nod++)
  {
    std::cout << " parents(" << nod << ") : " << ndag.getParents(nod)
              << std::endl;
    std::cout << "children(" << nod << ") : " << ndag.getChildren(nod)
              << std::endl;
  }

  std::cout << "topologicalOrder : " << dag.getTopologicalOrder() << std::endl;

  std::cout << std::endl << dag.__str__("__str__ |") << std::endl;

  std::cout << std::endl << dag.toDot();
}
void testConstructorWithErrors()
{
  std::cout << "\n\n** Checking raised exceptions " << std::endl;
  auto bn = gum::BayesNet<double>::fastPrototype("A->B->C<-D->E<-A");
  std::cout << "  - Before removing a node : " << bn << std::endl;
  bn.erase("C");
  std::cout << "  - After removing node C : " << bn << std::endl;

  std::cout << "=> Trying to build a dag with non contiguous ids :";
  try
  {
    OTAGRUM::NamedDAG d(bn);
    std::cout << " ERROR " << std::endl;
  }
  catch (OT::InvalidArgumentException)
  {
    std::cout << " OK " << std::endl;
  }
}

int main(void)
{
  testConstructor();
  testSerialisation();
  testConstructorWithErrors();

  return 0;
}

File: python/test/t_NamedDAG_std.py
#!/usr/bin/env python

import openturns.testing as ott
import otagrum
import pyagrum as gum


def testConstructorFromBayesNet():
    proto = "A->C->B->D;E->A->B<-E"
    bn = gum.BayesNet.fastPrototype(proto)
    ndag = otagrum.NamedDAG(bn)

    print("size : " + ("OK" if ndag.getSize() == 5 else "fail"))

    description = ndag.getDescription()
    idA = ndag.idFromName("A")
    idB = ndag.idFromName("B")
    idC = ndag.idFromName("C")
    idD = ndag.idFromName("D")
    idE = ndag.idFromName("E")

    print(
        "idFromName round-trips through description : "
        + ("OK" if description[idA] == "A" else "fail")
    )

    print("E has no parent : " + ("OK" if ndag.getParents(idE).getSize() == 0 else "fail"))
    print(
        "A has parent E : "
        + ("OK" if idE in list(ndag.getParents(idA)) else "fail")
    )
    print(
        "B has parents A, C and E : "
        + ("OK" if sorted(list(ndag.getParents(idB))) == sorted([idA, idC, idE]) else "fail")
    )
    print("D has no children : " + ("OK" if ndag.getChildren(idD).getSize() == 0 else "fail"))

    order = list(ndag.getTopologicalOrder())
    print("topological order has every node once : " + ("OK" if sorted(order) == list(range(5)) else "fail"))
    # A parent must appear before its children in the topological order.
    print(
        "topological order respects E before A : "
        + ("OK" if order.index(idE) < order.index(idA) else "fail")
    )


def testConstructorWithNonContiguousIdsRaises():
    bn = gum.BayesNet.fastPrototype("A->B->C<-D->E<-A")
    bn.erase("C")
    with ott.assert_raises(TypeError):
        otagrum.NamedDAG(bn)


testConstructorFromBayesNet()
testConstructorWithNonContiguousIdsRaises()
